internal/baggage/setter.go

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package baggage
+
+import (
+	"github.com/opentracing/opentracing-go"
+	"github.com/opentracing/opentracing-go/log"
+
+	"github.com/uber/jaeger-client-go"
+)
+
+type spanContext interface {
+	WithBaggageItem(key, value string) opentracing.SpanContext
+	IsSampled() bool
+	ForeachBaggageItem(handler func(k, v string) bool)
+}
+
+type span interface {
+	Context() opentracing.SpanContext
+	logFieldsNoLocking(fields ...log.Field)
```

Comment:
this is the main problem. Given that the function SetBaggage is run with the write lock acquired, if we want to write logs to the span we need to run this function. But given it's private we can't use it and duck typing doesn't work with private functions. The workaround would be to have the baggage setter return both the new SpanContext and the logs which ruins the abstraction.